### PR TITLE
Integrate FiftyOne for loading datasets and visualizing video classification predictions

### PR DESCRIPTION
**Pull request by ehofesmann**
## What does this PR do?

Integrates lightning flash with the analysis and visualization tool, FiftyOne. Primarily adds:
* `VideoClassificationData.from_fiftyone` loading data directly from FiftyOne datasets
* `FiftyOneLabels()` serializer returning predictions in the FiftyOne Classification(s) format

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/video/classification/data.py [hunk 1]
```diff
@@ -111,6 +127,84 @@ def predict_load_sample(self, sample: Dict[str, Any]) -> Dict[str, Any]:
...
+        self.clip_sampler = clip_sampler
+        self.video_sampler = video_sampler
+        self.decode_audio = decode_audio
+        self.decoder = decoder
+
+    def load_data(self, data: SampleCollection, dataset: Optional[Any] = None) -> 'EncodededVideoDataset':
+        label_to_class_mapping = dict(list(enumerate(data.default_classes)))
+        class_to_label_mapping = {c:l for l,c in label_to_class_mapping.items()}
+
+        labeled_video_paths = []
```
Comment (brimoor) [inline flash/video/classification/data.py]: This will be much more efficient when implemented this way:

```py
filepaths, labels = data.values(["filepath", self.label_field + ".label"])
targets = [class_to_label_mapping[l] for l in labels]
labeled_video_paths = list(zip(filepaths, targets))
```

File: flash/video/classification/data.py [hunk 2]
```diff
@@ -111,6 +127,84 @@ def predict_load_sample(self, sample: Dict[str, Any]) -> Dict[str, Any]:
         return self._encoded_video_to_dict(EncodedVideo.from_path(sample[DefaultDataKeys.INPUT]))
 
 
+class VideoClassificationFiftyOneDataSource(FiftyOneDataSource):
```
Comment (brimoor) [inline flash/video/classification/data.py]: This implementation seems to have quite a bit of overlap with `VideoClassificationPathsDataSource`. It would likely be cleaner to subclass from `VideoClassificationPathsDataSource` (and `FiftyOneDataSource`) and only reimplement the necessary bits

File: flash/core/classification.py [hunk 3]
```diff
@@ -158,3 +162,69 @@ def serialize(self, sample: Any) -> Union[int, List[int], str, List[str]]:
         else:
             rank_zero_warn("No LabelsState was found, this serializer will act as a Classes serializer.", UserWarning)
             return classes
+
+
+class FiftyOneLabels(Classes):
```
Comment (brimoor) [inline flash/core/classification.py]: You should subclass from `ClassificationSerializer` here because `Classes` is specifically for returning integer class label(s).

File: flash/core/data/data_module.py [hunk 4]
```diff
@@ -1024,3 +1030,77 @@ def from_datasets(
             num_workers=num_workers,
             **preprocess_kwargs,
         )
+
+    @classmethod
+    def from_fiftyone(
+        cls,
+        train_dataset: Optional[SampleCollection] = None,
```
Comment (brimoor) [inline flash/core/data/data_module.py:1037]: A point of discussion is whether to use `train_collection` or `train_dataset` terminology here, since, on one hand, people typically refer to "train dataset" and "test" dataset", while, for FiftyOne specifically, it is possible that both splits are merely views into a single `Dataset` instance.

For users familiar with FiftyOne, it will be meaningful to see "collection" instead of "dataset" because it means that they may provide any `SampleCollection`, not just a `Dataset`.

Comment (brimoor) [reply flash/core/data/data_module.py:1037]: Per live discussion, another point of consideration is whether we should allow a specification of the form `(dataset, train_tag, val_tag, test_tag)` as an alternative to the `(train_dataset, val_dataset, test_dataset)` format.

```py
data_module = VideoClassificationData.from_fiftyone(
    dataset=dataset,
    train_tag="train",
    val_tag="validation",
    test_tag="test",
    ...
)
```

Of course, if the user has data in "tagged dataset" format, then they could use the "separate dataset" syntax as follows:

```py
data_module = VideoClassificationData.from_fiftyone(
    train_dataset=dataset.match_tags("train"),
    val_dataset=dataset.match_tags("validation"),
    test_dataset=dataset.match_tags("test"),
    ...
)
```

File: flash/core/data/data_source.py [hunk 5]
```diff
@@ -461,3 +468,13 @@ class TensorDataSource(SequenceDataSource[torch.Tensor]):
 class NumpyDataSource(SequenceDataSource[np.ndarray]):
     """The ``NumpyDataSource`` is a ``SequenceDataSource`` which expects the input to
     :meth:`~flash.core.data.data_source.DataSource.load_data` to be a sequence of ``np.ndarray`` objects."""
+
+class FiftyOneDataSource(SequenceDataSource[SampleCollection]):
```
Comment (brimoor) [inline flash/core/data/data_source.py]: I believe the class signature here should be

```py
class FiftyOneDataSource(Generic[SampleCollection]):
```

The FiftyOne data source takes a `SampleCollection` as input, not a sequence of filepaths, numpy arrays, or anything like that.

They all *output* sequences, but the type structure refers to the *input* data.

File: flash/core/data/data_source.py [hunk 6]
```diff
@@ -461,3 +468,13 @@ class TensorDataSource(SequenceDataSource[torch.Tensor]):
...
+    """The ``FiftyOneDataSource`` is a ``SequenceDataSource`` which expects the input to
+    :meth:`~flash.core.data.data_source.DataSource.load_data` to be a sequence
+    of FiftyOne Dataset objects."""
+
+    def predict_load_data(self,
+                          data: SampleCollection,
+                          dataset: Optional[Any] = None) -> Sequence[Mapping[str, Any]]:
+        return [{DefaultDataKeys.INPUT: s.filepath} for s in data]
```
Comment (brimoor) [inline flash/core/data/data_source.py]: This will be more efficient:

```py
return [{DefaultDataKeys.INPUT: f} for f in data.values("filepath")]
```

File: flash/video/classification/data.py [hunk 7]
```diff
@@ -111,6 +127,84 @@ def predict_load_sample(self, sample: Dict[str, Any]) -> Dict[str, Any]:
...
+    ):
+        super().__init__()
+        self.label_field = label_field
+        self.clip_sampler = clip_sampler
+        self.video_sampler = video_sampler
+        self.decode_audio = decode_audio
+        self.decoder = decoder
+
+    def load_data(self, data: SampleCollection, dataset: Optional[Any] = None) -> 'EncodededVideoDataset':
+        label_to_class_mapping = dict(list(enumerate(data.default_classes)))
```
Comment (brimoor) [inline flash/video/classification/data.py]: `list()` is unnecessary

File: flash/core/data/data_module.py [hunk 8]
```diff
@@ -1024,3 +1030,77 @@ def from_datasets(
...
+        batch_size: int = 4,
+        num_workers: Optional[int] = None,
+        **preprocess_kwargs: Any,
+    ) -> 'DataModule':
+        """Creates a :class:`~flash.core.data.data_module.DataModule` object
+        from the given FiftyOne Datasets using the
+        :class:`~flash.core.data.data_source.DataSource` of name
+        :attr:`~flash.core.data.data_source.DefaultDataSources.FIFTYONE`
+        from the passed or constructed :class:`~flash.core.data.process.Preprocess`.
+        Args:
```
Comment (brimoor) [inline flash/core/data/data_module.py:1058]: add spaces between docstring sections

File: flash/video/classification/data.py [hunk 9]
```diff
@@ -119,6 +213,7 @@ def __init__(
         val_transform: Optional[Dict[str, Callable]] = None,
         test_transform: Optional[Dict[str, Callable]] = None,
         predict_transform: Optional[Dict[str, Callable]] = None,
+        label_field: str = "ground_truth",
```
Comment (brimoor) [inline flash/video/classification/data.py]: I bet the Flash team will object to a FiftyOne-specific parameter appearing here. It's probably best to replace `label_field` with `**kwargs` or ``**data_source_kwargs`` and document that these are source-specific keyword arguments that are passed to the `DataSource`.

File: flash/core/classification.py [hunk 10]
```diff
@@ -158,3 +162,69 @@ def serialize(self, sample: Any) -> Union[int, List[int], str, List[str]]:
...
+                    )
+                    classifications.append(fo_cls)
+                fo_labels = Classifications(
+                    classifications=classifications,
+                )
+            else:
+                fo_labels = Classification(
+                    label = labels[classes],
+                    confidence = max(probabilities),
+                    logits = sample.tolist(),
```
Comment (brimoor) [inline flash/core/classification.py]: I would recommend omitting the logits by default. If you want to provide support for storing logits (potentially *way* more storage), then do so via an optional kwarg that defaults to False.

File: flash/core/classification.py [hunk 11]
```diff
@@ -158,3 +162,69 @@ def serialize(self, sample: Any) -> Union[int, List[int], str, List[str]]:
...
+
+        classes = super().serialize(sample)
+
+        if self.multi_label:
+            probabilities = torch.sigmoid(sample).tolist()
+        else:
+            probabilities = torch.softmax(sample, -1).tolist()
+
+        if labels is not None:
+            if self.multi_label:
```
Comment (brimoor) [inline flash/core/classification.py:224]: the `threshold` needs to be used here for multilabel case

File: flash/core/classification.py [hunk 12]
```diff
@@ -158,3 +162,69 @@ def serialize(self, sample: Any) -> Union[int, List[int], str, List[str]]:
...
+                classifications = []
+                for probs, cls in zip(probabilities, classes):
+                    fo_cls = Classification(
+                        label = labels[cls],
+                        confidence = max(prob),
+                    )
+                    classifications.append(fo_cls)
+                fo_labels = Classifications(
```
Comment (brimoor) [inline flash/core/classification.py:232]: `Classifications` also provides a `logits` property, which is where logits should be stored in this case

**Review by brimoor — CHANGES_REQUESTED** [review]

File: flash/video/classification/data.py (hunk 2 above)
Comment (ehofesmann) [reply flash/video/classification/data.py]: I was starting to implement this, but then it seemed weird that it would then inherit from the `PathDataSource` and from the `FiftyOneDataSource` which seemed like it could cause issues since they are different `DataSource` subclasses

File: flash/video/classification/data.py (hunk 1 above)
Comment (ehofesmann) [reply flash/video/classification/data.py]: It seems like `data.values([...])` doesn't work at the moment, I'm just using `data.values("filepaths")` and `data.values(self.label_field + ".label")` instead

Comment (brimoor) [reply flash/video/classification/data.py]: The syntax I used was added a few days ago in https://github.com/voxel51/fiftyone/pull/1053 and should be available if you merge `develop` into this branch.
